pkg/apiserver/api_installer.go

```diff
@@ -110,96 +110,90 @@ func (a *APIInstaller) NewWebService() *restful.WebService {
 	return ws
 }
 
-func (a *APIInstaller) registerResourceHandlers(path string, storage rest.Storage, ws *restful.WebService, proxyHandler http.Handler) (*unversioned.APIResource, error) {
-	admit := a.group.Admit
-	context := a.group.Context
-
-	optionsExternalVersion := a.group.GroupVersion
-	if a.group.OptionsExternalVersion != nil {
-		optionsExternalVersion = *a.group.OptionsExternalVersion
-	}
-
-	var resource, subresource string
-	switch parts := strings.Split(path, "/"); len(parts) {
-	case 2:
-		resource, subresource = parts[0], parts[1]
-	case 1:
-		resource = parts[0]
-	default:
-		// TODO: support deeper paths
-		return nil, fmt.Errorf("api_installer allows only one or two segment paths (resource or resource/subresource)")
+// getResourceKind validates and returns the group version kind registered for the given
+// storage object. If the storage object is a subresource, it just return's the object's group
+// version kind without validating it's group against the group we are installing the REST
+// handlers for here.
+func (a *APIInstaller) getResourceKind(path string, storage rest.Storage) (unversioned.GroupVersionKind, error) {
+	if fqKindToRegister, ok := a.group.SubresourceGroupVersionKind[path]; ok {
+		return fqKindToRegister, nil
 	}
-	hasSubresource := len(subresource) > 0
 
 	object := storage.New()
 	fqKinds, err := a.group.Typer.ObjectKinds(object)
 	if err != nil {
-		return nil, err
+		return unversioned.GroupVersionKind{}, err
 	}
+
 	// a given go type can have multiple potential fully qualified kinds.  Find the one that corresponds with the group
 	// we're trying to register here
 	fqKindToRegister := unversioned.GroupVersionKind{}
 	for _, fqKind := range fqKinds {
 		if fqKind.Group == a.group.GroupVersion.Group {
-			fqKindToRegister = fqKind
+			fqKindToRegister = a.group.GroupVersion.WithKind(fqKind.Kind)
 			break
 		}
 
 		// TODO This keeps it doing what it was doing before, but it doesn't feel right.
 		if fqKind.Group == extensions.GroupName && fqKind.Kind == "ThirdPartyResourceData" {
-			fqKindToRegister = fqKind
-			fqKindToRegister.Group = a.group.GroupVersion.Group
-			fqKindToRegister.Version = a.group.GroupVersion.Version
+			fqKindToRegister = a.group.GroupVersion.WithKind(fqKind.Kind)
 		}
 	}
+	if fqKindToRegister.IsEmpty() {
+		return unversioned.GroupVersionKind{}, fmt.Errorf("unable to locate fully qualified kind for %v: found %v when registering for %v", reflect.TypeOf(object), fqKinds, a.group.GroupVersion)
+	}
+	return fqKindToRegister, nil
+}
 
-	kind := fqKindToRegister.Kind
+// restMapping returns the RESTMapping object for the given resource or the subresource based on
+// the supplied subresource overrides.
+func (a *APIInstaller) restMapping(path, resource string, hasSubresource bool, fqKindToRegister unversioned.GroupVersionKind) (*meta.RESTMapping, error) {
```

Comment:
It's extremely strange that this function expects fqKindToRegister to not correspond to resource.

Like maybe just don't take path, hasSubresource or fqKindToRegister as parameters and always do the section in the `if`.